sklearn/neighbors/quad_tree.pyx

```diff
@@ -0,0 +1,547 @@
+        cdef DTYPE_t[3] pt
+        cdef DTYPE_t[3] min_bounds, max_bounds
+
+        # validate X and prepare for query
+        # X = check_array(X, dtype=DTYPE_t, order='C')
+        n_samples = X.shape[0]
+
+        capacity = 100
+        self._resize(capacity)
+        m, M = np.min(X, axis=0) - 1e-3, np.max(X, axis=0) + 1e-3
+        for i in range(self.n_dimensions):
+            min_bounds[i] = m[i]
+            max_bounds[i] = M[i]
+
+        # Create the initial node with boundaries from the dataset
+        self._init_root(min_bounds, max_bounds)
+
+        for i in range(n_samples):
+            for j in range(self.n_dimensions):
+                pt[j] = X[i, j]
+            self.insert_point(pt, i)
+
+        self._resize(capacity=self.cell_count)
+
+    def plot_tree(self):
+        """Plot the tree with cell boundaries and the points inserted in it."""
+        self.check_coherence()
+        import matplotlib.pyplot as plt
+
+        plt.figure()
+        for c in self.cells[:self.cell_count]:
+            if not c.is_leaf:
+                # Plot the cell division if the cell is an inner cell
+                plt.vlines(c.center[0], c.min_bounds[1], c.max_bounds[1])
+                plt.hlines(c.center[1], c.min_bounds[0], c.max_bounds[0])
+            else:
+                # If the cell is a leaf, display the point contained in it.
+                plt.scatter(c.barycenter[0], c.barycenter[1], c='b', marker='.')
+            
+        # Print bounding box of the Tree
+        root = self.cells[0]
+        plt.vlines([root.min_bounds[0], root.max_bounds[0]], root.min_bounds[1], root.max_bounds[1])
+        plt.hlines([root.min_bounds[1], root.max_bounds[1]], root.min_bounds[0], root.max_bounds[0])
+        plt.show()
+
+    def check_coherence(self):
+        for cell in self.cells[:self.cell_count]:
+            self.check_point_in_cell(cell.barycenter, &cell)
+            if not cell.is_leaf:
+                n_points = 0
+                for idx in range(self.n_cells_per_cell):
+                    child_id = cell.children[idx]
+                    if child_id != -1:
+                        child = self.cells[child_id]
+                        n_points += child.cumulative_size
+                        assert child.cell_id == child_id, (
+                            "Cell id not correctly initiliazed.")
+                if n_points != cell.cumulative_size:
+                    raise RuntimeError(
+                        "Cell {} is incoherent. Size={} but found {} points "
+                        "in children. ({})"
+                        .format(cell.cell_id, cell.cumulative_size,
+                                n_points, cell.children))
+        if self.n_points != self.cells[0].cumulative_size:
+            raise RuntimeError(
+                "QuadTree is incoherent. Size={} but found {} points "
+                "in children."
+                .format(self.n_points, self.cells[0].cumulative_size))
```

Comment:
This function is used to verify that the different information computed in the different `Cell` are compatible with the info in the other `Cell`